FlowCrypt/src/main/java/com/flowcrypt/email/jetpack/viewmodel/RefreshPrivateKeysFromEkmViewModel.kt

```diff
@@ -0,0 +1,233 @@
+ */
+class RefreshPrivateKeysFromEkmViewModel(application: Application) : AccountViewModel(application) {
+  private val repository = FlowcryptApiRepository()
+  private val controlledRunnerRefreshPrivateKeysFromEkm = ControlledRunner<Result<Boolean?>>()
+  private val refreshPrivateKeysFromEkmMutableStateFlow =
+    MutableStateFlow<Result<Boolean?>>(Result.none())
+  val refreshPrivateKeysFromEkmStateFlow =
+    refreshPrivateKeysFromEkmMutableStateFlow.asStateFlow()
+
+  fun refreshPrivateKeys() {
+    viewModelScope.launch {
+      refreshPrivateKeysFromEkmMutableStateFlow.value = Result.loading()
+      val activeAccount = getActiveAccountSuspend() ?: return@launch
+      if (activeAccount.clientConfiguration?.usesKeyManager() == true) {
+        refreshPrivateKeysFromEkmMutableStateFlow.value =
+          controlledRunnerRefreshPrivateKeysFromEkm.joinPreviousOrRun {
+            return@joinPreviousOrRun try {
+              refreshPrivateKeysInternally(activeAccount)
+            } catch (e: Exception) {
+              e.printStackTrace()
+              Result.exception(e)
+            }
+          }
+      } else {
+        refreshPrivateKeysFromEkmMutableStateFlow.value = Result.success(true)
+      }
+    }
+  }
+
+  private suspend fun refreshPrivateKeysInternally(activeAccount: AccountEntity): Result<Boolean?> =
+    withContext(Dispatchers.IO) {
+      val context: Context = getApplication()
+      val retryAttempts = 6
+      val idToken = GeneralUtil.getGoogleIdToken(
+        context = context,
+        maxRetryAttemptCount = retryAttempts
+      )
+
+      val ekmPrivateResult = repository.getPrivateKeysViaEkm(
+        context = context,
+        ekmUrl = requireNotNull(activeAccount.clientConfiguration?.keyManagerUrl),
+        idToken = idToken
+      )
+
+      if (ekmPrivateResult.status != Result.Status.SUCCESS) {
+        return@withContext when (ekmPrivateResult.status) {
+          Result.Status.EXCEPTION -> {
+            Result.exception(
+              //Thrown network errors can be ignored but not other types of errors
+              GeneralUtil.preProcessException(
+                context = context,
+                causedException = requireNotNull(ekmPrivateResult.exception)
+              )
+            )
+          }
+
+          Result.Status.ERROR -> Result.exception(ApiException(ekmPrivateResult.data?.apiError))
+
+          else -> throw IllegalStateException(
+            "Unsupported status = ${ekmPrivateResult.status} at this step"
+          )
+        }
+      }
+
+      if (ekmPrivateResult.data?.privateKeys?.isEmpty() == true) {
+        return@withContext Result.success(true)
+      }
+
+      val pgpKeyDetailsList = mutableListOf<PgpKeyDetails>()
+      ekmPrivateResult.data?.privateKeys?.forEach { key ->
+        val parsedList = PgpKey.parsePrivateKeys(requireNotNull(key.decryptedPrivateKey))
+          .map {
+            it.copy(
+              passphraseType = KeyEntity.PassphraseType.RAM,
```

Comment:
I suppose the PassphraseType should follow whichever client configuration the user has set up, to reflect the pass phrase types of other keys.

Reply:
It's a little difficult question. `passphraseType` can be different for existing keys(the common logic). However, for enterprise users during the setup process, we mark private keys that were received from EKM as `passphraseType = KeyEntity.PassphraseType.RAM`. If rules don't allow to import external keys there is no chance to have other `passphraseType`.

But if we will have a situation where will be possible to import external keys I will not know what `passphraseType` to use for a new key.

For now I don't change `passphraseType` for existing keys(that will be updated) and use `passphraseType = KeyEntity.PassphraseType.RAM` for new keys.

Reply:
Is setup with EKM also allowed when `FORBID_STORING_PASS_PHRASE` is not set?

Theoretically, it would be (it is on browser extension) but we disallowed this on Android?

Reply:
> Is setup with EKM also allowed when FORBID_STORING_PASS_PHRASE is not set?

No, we disallow it.

https://github.com/FlowCrypt/flowcrypt-android/blob/5bb839f01b6780c1cd8d4d454b0e81e232e8b0c7/FlowCrypt/src/main/java/com/flowcrypt/email/jetpack/viewmodel/EkmViewModel.kt#L117

respectively to https://github.com/FlowCrypt/flowcrypt-android/issues/1168

Reply:
> app auto updates keys from EKM during startup without pass phrase prompt wasn't implemented?

Why do you think so?

After start the app we

1. https://github.com/FlowCrypt/flowcrypt-android/blob/9f47a4797783a59538984d1f5cdd80e778d05f8d/FlowCrypt/src/main/java/com/flowcrypt/email/ui/activity/MainActivity.kt#L423
2. https://github.com/FlowCrypt/flowcrypt-android/blob/9f47a4797783a59538984d1f5cdd80e778d05f8d/FlowCrypt/src/main/java/com/flowcrypt/email/jetpack/viewmodel/RefreshPrivateKeysFromEkmViewModel.kt#L145
3. https://github.com/FlowCrypt/flowcrypt-android/blob/9f47a4797783a59538984d1f5cdd80e778d05f8d/FlowCrypt/src/main/java/com/flowcrypt/email/jetpack/viewmodel/RefreshPrivateKeysFromEkmViewModel.kt#L181
4. or https://github.com/FlowCrypt/flowcrypt-android/blob/9f47a4797783a59538984d1f5cdd80e778d05f8d/FlowCrypt/src/main/java/com/flowcrypt/email/jetpack/viewmodel/RefreshPrivateKeysFromEkmViewModel.kt#L198
5. and https://github.com/FlowCrypt/flowcrypt-android/blob/9f47a4797783a59538984d1f5cdd80e778d05f8d/FlowCrypt/src/main/java/com/flowcrypt/email/jetpack/viewmodel/RefreshPrivateKeysFromEkmViewModel.kt#L147 that means success

Reply:
I see. It's implemented, but this scenario cannot be effectively tried because this scenario is not allowed yet during EKM setup. Is that correct?

Reply:
> during EKM setup

Do you mean the setup account process?

> Is that correct?

I think I'm starting to understand what you mean. Due to #1765, we can't cover all cases `for app auto updates keys from EKM during startup without pass phrase prompt`. Because we can have a situation where a user has a key with pass phrase in the local database. Respectively the current logic we save a new key with `passphraseType = KeyEntity.PassphraseType.RAM` always, independently to the 'passphraseType' of the used pass phrase.

Reply:
As implemented, yes. That's why I highlighted it in #1765, so we know about it whenever we remove that setup limitation. Does that make sense?

Reply:
> Does that make sense?

I think if we depend on `OrgRules` on the setup account stage we have to follow the same logic here. if `FORBID_STORING_PASS_PHRASE` is used we should use here pass phrases with `passphraseType =KeyEntity.PassphraseType.RAM` only.

To be honest, there can be a lot of nuances. For example, we can have `FORBID_STORING_PASS_PHRASE` during the setup process, aftert some time rules can be updated and `FORBID_STORING_PASS_PHRASE` will be deleted, after that a user will import a key with `passphraseType =KeyEntity.PassphraseType.DATABASE`, after that rules can be updated and `FORBID_STORING_PASS_PHRASE` will be added(but what about keys with `passphraseType =KeyEntity.PassphraseType.DATABASE`) and so on...

> I think if we depend on `OrgRules` on the setup account stage we have to follow the same logic here. if `FORBID_STORING_PASS_PHRASE` is used we should use here pass phrases with `passphraseType =KeyEntity.PassphraseType.RAM` only.

But what about the case if we will not have such a pass phrase? how to handle it?

Reply:
I think this can be discussed and addressed in #1765 since this scenario is not supported during setup. Therefore it shouldn't happen here.